### PR TITLE
TST: follow numpy/pytest traceback conventions in simple assert_* test helper functions

### LINKED ISSUE #16071: TST: should we adopt `__tracebackhide__=True` in `assert_*` test helper functions ?
`numpy.testing.assert_*` functions set `__traceback__ = True`, which changes `pytest`'s behavior so that the internals of these functions is not shown in errors. This is particularly useful when invoking `pytest` with `--pdb`, which opens a debugger when (and *where*) an exception is raised: because the internals of `assert_*` functions are excluded from the traceback, the *location* associated with the error is that of the *callsite*, which is usually the body of a test function.
Let me illustrate what I mean with a simplified example

```python
# t.py
import numpy as np

def assert_equal_len_v1(x, y):
    assert len(x) == len(y)

def assert_equal_len_v2(x, y):
    __tracebackhide__ = True
    assert len(x) == len(y)

def test_v1():
    a = np.arange(2)
    b = np.arange(3)
    assert_equal_len_v1(a, b)

def test_v2():
    a = np.arange(2)
    b = np.arange(3)
    assert_equal_len_v2(a, b)
```

run with `pytest --pdb t.py` and see *where* the debugger is opened: `test_v1` will open in within `assert_equal_v1`'s body, while `test_v2` opens it in its own body, which is arguably more useful since we can inspect the whole local scope, instead of just the couple variables that were passed (under different names !) to a test helper.

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description

Fixes #16071 and replaces #16069

Reviews:
- [ ] convolution - @keflavich 
- [x] coordinates
- [x] io.ascii
- [ ] io.fits - @saimn 
- ~io.misc~
- [x] io.votable
- [x] nddata - @mwcraig 
- [x] table
- [x] time
- [ ] timeseries - @astrofrog 
- [x] uncertainty - @mhvk 
- [x] units
- [x] utils.masked - @mhvk 

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/io/fits/tests/test_connect.py
Comment (saimn) [inline astropy/io/fits/tests/test_connect.py:845]: Not sure to understand the benefit. If a test fail I will want to know why and hence dive into this function to find which attribute and examine the value, so why hiding it ?

Comment (neutrinoceros) [reply astropy/io/fits/tests/test_connect.py:845]: I tried to explain the expected gain in #16080
I could furthermore add that:
- in a sense, this satisfies the principle of least surprise (as a user of `numpy.testing.assert_*` functions, I grew to expect the internals of such functions to be hidden in tracebacks)
- you can still opt-in the current behaviour using `--full-trace`, while the behaviour this is enabling is *not* achievable currently

Comment (WilliamJamieson) [reply astropy/io/fits/tests/test_connect.py:845]: I personally would be more surprised by not being able to know why I had a failure. The traceback would still go through the failure within the unit test. Often I end up having to bisect issues by examining these helper functions and making temporary modifications to them.

If we do go through with this change I suggest we carefully consider the error messages being raised so that they are more indicative of what has caused the failure.

Comment (saimn) [reply astropy/io/fits/tests/test_connect.py:845]: Yes but being useful for `numpy.testing.assert_` doesn't mean it's always better. Here the code is specific to what is being tested and used by a few tests only, and there is a good chance that I will want to see the full traceback is there a problem.
Also the problem is that I don't know what to expect with `__tracebackhide__` activated, what will pytest print ?

File: astropy/io/ascii/tests/common.py
Comment (mhvk) [inline astropy/io/ascii/tests/common.py:27]: Hmm, really these should just be replaced! But can be for follow-up...

File: astropy/io/ascii/tests/test_ecsv.py
Comment (mhvk) [inline astropy/io/ascii/tests/test_ecsv.py:295]: Here I am much less sure, since really this is a combination of tests; I think in debugging it would actually be useful to know which part failed.

File: astropy/io/ascii/tests/test_rst.py
Comment (mhvk) [inline astropy/io/ascii/tests/test_rst.py:14]: The same as in the test above... since even in the same module, just import it?

File: astropy/io/fits/tests/test_connect.py
Comment (mhvk) [reply astropy/io/fits/tests/test_connect.py:845]: As you'll have seen from my comment above, I agree with @saimn here - I can see representations as being a single object where one wants to hide the traceback, since their content is uniform, but not a more complicated object where one also compares attributes.

File: astropy/io/misc/tests/test_hdf5.py
Comment (mhvk) [inline astropy/io/misc/tests/test_hdf5.py:668]: Same again.

Note that we really should refactor to ensure these very similar functions are not repeated in multiple places!

File: astropy/io/misc/tests/test_parquet.py
Comment (mhvk) [inline astropy/io/misc/tests/test_parquet.py:549]: Another one, both for not hidiing and for refactoring...

File: astropy/samp/tests/test_helpers.py
Comment (mhvk) [inline astropy/samp/tests/test_helpers.py:23]: This also feels too complicated for hiding the trace...

File: astropy/tests/helper.py
Comment (mhvk) [inline astropy/tests/helper.py:90]: Here again one would want to know where things failed.

File: astropy/units/tests/test_quantity_info.py
Comment (mhvk) [inline astropy/units/tests/test_quantity_info.py:12]: Similarly, not in favour of this one.

File: astropy/utils/masked/tests/test_masked.py
Comment (mhvk) [inline astropy/utils/masked/tests/test_masked.py:21]: This one is fine.

File: astropy/visualization/wcsaxes/tests/test_coordinate_helpers.py
Comment (mhvk) [inline astropy/visualization/wcsaxes/tests/test_coordinate_helpers.py:41]: I wouldn't do this one.

File: astropy/wcs/wcsapi/tests/test_fitswcs.py
Comment (mhvk) [inline astropy/wcs/wcsapi/tests/test_fitswcs.py:598]: Not this one either.

**Review by mhvk — COMMENTED** [review]
While I like the general idea, I think this goes too far. Note that the numpy `assert_array_equal`, etc., actually tell you about the mismatched elements; they don't just fail. I think in such a case there is indeed no reason to keep the full trace. However, absent that, one should have the trace for any function which does not have such a message - so I think we should not change any of the more complicated functions, where a lot of different things can go wrong.

Anyway, bottom line, I'd only do the simple ones that use `assert_array_equal` etc under the hood.

File: astropy/io/ascii/tests/common.py
Comment (neutrinoceros) [reply astropy/io/ascii/tests/common.py:27]: Right. I can do that as a follow up !

File: astropy/io/ascii/tests/test_ecsv.py
Comment (neutrinoceros) [reply astropy/io/ascii/tests/test_ecsv.py:295]: good thinking, I'll clean it up.

File: astropy/io/misc/tests/test_hdf5.py
Comment (neutrinoceros) [reply astropy/io/misc/tests/test_hdf5.py:668]: this one in particular has at least 2 variants (both are copy-pasted at least once). All of them are in the `io` namespace, maybe we should move them to a unified (private) test helper module (`astropy.io._testing`) ?

**Comment by neutrinoceros** [comment]
Sorry folks, I realize I made a mistake by trying to do this as quickly as possible instead of going through a careful review myself. I didn't want to waste contract money but I ended up wasting reviewer time, which is probably worse.
Anyway, I've reverted the most controversial parts of the patch so this now only applies to *simple* `assert_*` functions. I've also pushed one very small refactor (deduplicated `assert_equal_splitlines`, which is a 1 line function), and I took note of other suggested refactors for follow ups. Namely:
- inline functions from `io.ascii.tests.common` (`assert_true`, `assert_almost_equal` and `assert_equal`)
- deduplicate `def assert_objects_equal(obj1, obj2, attrs, compare_class=True)` (currently 4 matches, with 2 variants)

File: astropy/io/misc/tests/test_parquet.py
Comment (neutrinoceros) [reply astropy/io/misc/tests/test_parquet.py]: I'll keep the refactoring for a follow up if that's alright, but I'll revert this change

File: astropy/io/fits/tests/test_connect.py
Comment (neutrinoceros) [reply astropy/io/fits/tests/test_connect.py]: > If we do go through with this change I suggest we carefully consider the error messages being raised so that they are more indicative of what has caused the failure.

> Also the problem is that I don't know what to expect with __tracebackhide__ activated, what will pytest print ?

In short: pytest prints the exact same message as it would without `__tracebackhide__`, the difference is in the line that is associated with that message.

To illustrate let's take my `t.py` test module from #16071 (copied here in a dropdown for convenience)
<details><summary>t.py</summary>

```python
# t.py
import numpy as np

def assert_equal_len_v1(x, y):
    assert len(x) == len(y)

def assert_equal_len_v2(x, y):
    __tracebackhide__ = True
    assert len(x) == len(y)

def test_v1():
    a = np.arange(2)
    b = np.arange(3)
    assert_equal_len_v1(a, b)

def test_v2():
    a = np.arange(2)
    b = np.arange(3)
    assert_equal_len_v2(a, b)
```

</details>

Here's the output
```
$ pytest t.py
====================== test session starts ======================
platform darwin -- Python 3.12.0, pytest-8.0.0, pluggy-1.4.0
...
t.py FF                                                   [100%]

=========================== FAILURES ============================
____________________________ test_v1 ____________________________

    def test_v1():
        a = np.arange(2)
        b = np.arange(3)
>       assert_equal_len_v1(a, b)

t.py:14:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = array([0, 1]), y = array([0, 1, 2])

    def assert_equal_len_v1(x, y):
>       assert len(x) == len(y)
E       assert 2 == 3
E        +  where 2 = len(array([0, 1]))
E        +  and   3 = len(array([0, 1, 2]))

t.py:5: AssertionError
____________________________ test_v2 ____________________________

    def test_v2():
        a = np.arange(2)
        b = np.arange(3)
>       assert_equal_len_v2(a, b)
E       assert 2 == 3
E        +  where 2 = len(array([0, 1]))
E        +  and   3 = len(array([0, 1, 2]))

t.py:19: AssertionError
==================== short test summary info ====================
FAILED t.py::test_v1 - assert 2 == 3
FAILED t.py::test_v2 - assert 2 == 3
======================= 2 failed in 0.06s =======================
```

**Review by taldcroft — APPROVED** [review]
Fine for table, time, io.ascii

File: astropy/io/misc/tests/test_hdf5.py
Comment (mhvk) [reply astropy/io/misc/tests/test_hdf5.py]: `io.tests` already contains some common stuff (`mixin_columns.py`); it seems fine to add more things there (though perhaps not in this PR!).

Comment (neutrinoceros) [reply astropy/io/misc/tests/test_hdf5.py]: noted !

File: astropy/io/ascii/tests/common.py
Comment (eerovaher) [reply astropy/io/ascii/tests/common.py]: There's an open issue for this already: #15782

Comment (dhomeier) [reply astropy/io/ascii/tests/common.py]: These are being removed already in #15839, but in most instances the corresponding `numpy.testing` functions would be used; so if I understood you correctly, the numpy versions would already take care of setting `__tracebackhide__`? A couple uses of `assert_equal` would be changed to simple `assert`s though.

Comment (neutrinoceros) [reply astropy/io/ascii/tests/common.py]: thanks @eerovaher , I wasn't aware of this issue. Since it's mark with "good first issue" I should leave it to new contributors, right ?

> if I understood you correctly, the numpy versions would already take care of setting `__tracebackhide__`?

Correct ! For instance, see the first line in `numpy.testing.assert_equal`'s definition https://github.com/numpy/numpy/blob/e920ca42d98c4911c2294784f8809b38735923a4/numpy/testing/_private/utils.py#L313

Comment (eerovaher) [reply astropy/io/ascii/tests/common.py]: Simple and non-critical tasks such as removing these functions are indeed best left for new contributors.

Comment (dhomeier) [reply astropy/io/ascii/tests/common.py]: > thanks @eerovaher , I wasn't aware of this issue. Since it's mark with "good first issue" I should leave it to new contributors, right ?
> 
It is all but done in #15839 already, but your comments would be very welcome there. Pretty much all that remains to be done is settling on where to best use `numpy.testing.assert_equal()` vs. `assert a == b`. And if this PR is being merged first, might be worth looking if the new `assert`s in #15839 should get their `__tracebackhide__` as well.

Comment (neutrinoceros) [reply astropy/io/ascii/tests/common.py]: Thank you, I'll make sure to review this one tomorrow !

File: astropy/io/ascii/tests/common.py
Comment (dhomeier) [inline astropy/io/ascii/tests/common.py:42]:
```suggestion
    __tracebackhide__ = True
    assert arg1.splitlines() == arg2.splitlines()
```
As the above will cause problems once `assert_equal` is removed in #15839 might be best to inline that function already now, or alternatively use `np.testing.assert_equal` here, I guess (obviating the need for an extra `__tracebackhide__`). Again hard to say which one is better in terms of a more useful traceback; since `assert_equal_splitlines` is mostly used on small tables here with not all columns numerical, I'd tend to the simple `assert`.

**Comment by neutrinoceros** [comment]
I'm not sure all labels are still appropriate, they were added by the bot based on my initial commit, which had a larger scope.
In particular `samp`, `visualization.wcsaxes` and `wcs.wcsapi` seem outdated.
But that's still a lot of different packages touched in one PR, so I'm happy to split everything that was already approved parts into an isolated PR if desired.

**Comment by pllim** [comment]
Updated the labels. FYI.

Yes, the bot does not re-label every commit (and I think doing that would be a little excessive).

**Review by nstarman — APPROVED** [review]

**Comment by pllim** [comment]
@saimn , how do you feel about `io.fits`?

**Review by eerovaher — CHANGES_REQUESTED** [review]
Setting `__tracebackhide__ = True` is a good idea if the test helper function is written to produce good output in case of failure (like the `numpy.testing` functions are) or if it is a thin wrapper around such a function. It looks like in `coordinates` the only function that should have `__tracebackhide__ = True` is `assert_representation_allclose()` in `test_representation_arithmetic.py`.

**Comment by neutrinoceros** [comment]
@eerovaher thank you, I've reverted everything else in `coordinates`

**Review by eerovaher — APPROVED** [review]
I can approve for `coordinates`, assuming 5d8430b4d6cec81458057dcc442d1262180d33f5 gets squashed away.

**Comment by neutrinoceros** [comment]
(squashed)

**Comment by neutrinoceros** [comment]
rebased to fix a trivial conflict. I also added a checklist to keep track of reviews, if we absolutely require every package maintainer to approve.

**Review by pllim — APPROVED** [review]
Does not hurt for `io.votable`.

File: astropy/io/misc/tests/test_parquet.py
Comment (pllim) [inline astropy/io/misc/tests/test_parquet.py:1]: Not sure why this diff is here.

Comment (neutrinoceros) [reply astropy/io/misc/tests/test_parquet.py:1]: good catch. That's a left over from a previous rebased, I removed it now !

**Review by mhvk — APPROVED** [review]
utils.masked and uncertainties are OK; thanks for pushing this forward!

**Review by mwcraig — APPROVED** [review]

**Comment by mhvk** [comment]
I think this is really all fine - and a nice addition - so let's get it in. Thanks, @neutrinoceros!

**Pull request merged**
